File: lib/runtime/actorRoutineQueue.ts
export interface ActorRoutineContext {
  readonly actorId: number
  readonly targetId?: number
}

export interface RoutineHandle {
  hasCompletedAllSequences(): boolean
}

export class RoutineOptions {
  readonly blocking: boolean
  readonly highPriority: boolean
  readonly expiryDurationFrames: number | null
  readonly callback: (() => void) | null

  constructor(options: {
    blocking?: boolean
    highPriority?: boolean
    expiryDurationFrames?: number | null
    callback?: (() => void) | null
  } = {}) {
    this.blocking = options.blocking ?? true
    this.highPriority = options.highPriority ?? false
    this.expiryDurationFrames = options.expiryDurationFrames ?? null
    this.callback = options.callback ?? null
  }
}

interface RoutineProvider<TRoutine> {
  ready: () => boolean
  provider: () => TRoutine | null
}

interface QueuedRoutine<TRoutine> {
  readonly context: ActorRoutineContext
  readonly routineProvider: RoutineProvider<TRoutine>
  readonly options: RoutineOptions
  age: number
}

export class ActorRoutineQueue<TRoutine> {
  private readonly notReadyExpiryFrames: number
  private readonly isAnimationLocked: () => boolean
  private readonly playRoutine: (context: ActorRoutineContext, routine: TRoutine, options: RoutineOptions) => RoutineHandle | null

  private readonly routineQueue: QueuedRoutine<TRoutine>[] = []
  private currentRoutine: RoutineHandle | null = null

  constructor(options: {
    isAnimationLocked: () => boolean
    playRoutine: (context: ActorRoutineContext, routine: TRoutine, options: RoutineOptions) => RoutineHandle | null
    notReadyExpiryFrames?: number
  }) {
    this.isAnimationLocked = options.isAnimationLocked
    this.playRoutine = options.playRoutine
    this.notReadyExpiryFrames = options.notReadyExpiryFrames ?? 300
  }

  update(elapsedFrames: number): void {
    if (this.currentRoutine?.hasCompletedAllSequences()) {
      this.currentRoutine = null
    }

    this.routineQueue.forEach((queuedRoutine) => {
      queuedRoutine.age += elapsedFrames
    })

    this.removeExpiredEntries()

    const next = this.routineQueue[0]
    if (!next) {
      return
    }

    if (!next.routineProvider.ready()) {
      if (next.age >= this.notReadyExpiryFrames) {
        this.routineQueue.shift()
        this.executeCallback(next)
      }
      return
    }

    if (!next.options.highPriority && this.isAnimationLocked()) {
      return
    }

    this.routineQueue.shift()
    const routine = next.routineProvider.provider()
    if (!routine) {
      this.executeCallback(next)
      return
    }

    const handle = this.playRoutine(next.context, routine, next.options)
    if (next.options.blocking) {
      this.currentRoutine = handle
    } else {
      this.update(0)
    }
  }

  hasEnqueuedRoutines(): boolean {
    return this.currentRoutine !== null || this.routineQueue.length > 0
  }

  enqueueRoutine(context: ActorRoutineContext, routine: TRoutine | null, options = new RoutineOptions()): void {
    const wrapped: RoutineProvider<TRoutine> = {
      ready: () => true,
      provider: () => routine,
    }

    this.enqueueInternal({
      context,
      routineProvider: wrapped,
      options,
      age: 0,
    })
  }

  enqueueProvider(context: ActorRoutineContext, provider: () => TRoutine | null, options = new RoutineOptions()): void {
    const wrapped: RoutineProvider<TRoutine> = {
      ready: () => provider() !== null,
      provider,
    }
    this.enqueueInternal({
      context,
      routineProvider: wrapped,
      options,
      age: 0,
    })
  }

  private enqueueInternal(queued: QueuedRoutine<TRoutine>): void {
    if (!queued.options.highPriority) {
      this.routineQueue.push(queued)
      return
    }

    const index = this.routineQueue.findIndex((entry) => !entry.options.highPriority)
    if (index === -1) {
      this.routineQueue.push(queued)
    } else {
      this.routineQueue.splice(index, 0, queued)
    }
  }

  private removeExpiredEntries(): void {
    while (this.routineQueue.length > 0) {
      const next = this.routineQueue[0]
      if (!next || next.options.expiryDurationFrames === null || next.age < next.options.expiryDurationFrames) {
        return
      }

      this.routineQueue.shift()
      this.executeCallback(next)
    }
  }

  private executeCallback(queued: QueuedRoutine<TRoutine>): void {
    queued.options.callback?.()
  }
}

File: xim/poc/ActorRoutineQueue.kt
package xim.poc

import xim.poc.game.AttackContext
import xim.poc.game.AttackContexts
import xim.resource.*
import xim.util.Fps
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

class RoutineOptions(
    val blocking: Boolean = true,
    val highPriority: Boolean = false,
    val expiryDuration: Duration? = null,
    val playbackRate: Float = 1f,
    val callback: RoutineCompleteCallback? = null,
)

private class RoutineProvider(val ready: () -> Boolean, val provider: () -> EffectRoutineResource?) {

    fun isReady(): Boolean {
        return ready.invoke()
    }

    fun getIfPresent(): EffectRoutineResource? {
        return provider.invoke()
    }

}

private class QueuedRoutine(
    val context: ActorContext,
    val routineProvider: RoutineProvider,
    val options: RoutineOptions = RoutineOptions(),
) {
    var age = 0f
}

class ActorRoutineQueue(private val actor: Actor) {

    private val notReadyExpiryDuration = 5.seconds

    private val routineQueue = ArrayDeque<QueuedRoutine>()
    private var currentRoutine: EffectRoutineInstance? = null

    fun update(elapsedFrames: Float) {
        val actorModel = actor.actorModel ?: return

        if (currentRoutine?.hasCompletedAllSequences() == true) { currentRoutine = null }

        routineQueue.forEach { it.age += elapsedFrames }
        removeExpiredEntries()

        val next = routineQueue.firstOrNull() ?: return
        if (!next.routineProvider.isReady()) {
            removeNotReadyRoutineIfNeeded(next)
            return
        }

        if (!next.options.highPriority && actorModel.isAnimationLocked()) { return }

        routineQueue.removeFirst()

        val routine = next.routineProvider.getIfPresent()
        if (routine == null) {
            executeCallbacks(next)
            return
        }

        val routineInstance = EffectManager.registerActorRoutine(actor, next.context, routine)
        routineInstance.registerAllEffectsCompletedCallback(next.options.callback)
        routineInstance.animationPlaybackRate = next.options.playbackRate

        if (next.options.blocking) { currentRoutine = routineInstance } else { update(0f) }
    }

    fun hasEnqueuedRoutines(): Boolean {
        return currentRoutine != null || routineQueue.isNotEmpty()
    }

    fun enqueueRoutine(context: ActorContext, options: RoutineOptions, effectRoutineResource: EffectRoutineResource?) {
        val provider = RoutineProvider(ready = { true }, provider = { effectRoutineResource })
        enqueueRoutine(context, provider, options)
    }

    fun enqueueRoutine(context: ActorContext, options: RoutineOptions, routineProvider: () -> EffectRoutineResource?) {
        val provider = RoutineProvider(ready = { routineProvider.invoke() != null }, provider = { routineProvider.invoke() })
        enqueueRoutine(context, provider, options)
    }

    fun enqueueMainBattleRoutine(datId: DatId, targetId: ActorId, attackContext: AttackContext = AttackContext(), callback: (() -> Unit)? = null) {
        enqueueMainBattleRoutine(datId = datId, attackContext = attackContext, targetId = targetId, options = RoutineOptions(blocking = false, callback = callback))
    }

    fun enqueueMainBattleRoutine(datId: DatId, targetId: ActorId, attackContext: AttackContext, options: RoutineOptions) {
        val context = ActorContext(actor.id, targetId, attackContexts = AttackContexts.single(targetId, attackContext))

        enqueueRoutine(datId, context = context, options = options) {
            actor.actorModel?.model?.getMainBattleAnimationDirectory()
        }
    }

    fun enqueueMainBattleRoutine(targetId: ActorId, attackContext: AttackContext, options: RoutineOptions, animProvider: () -> DatId) {
        val context = ActorContext(actor.id, targetId, attackContexts = AttackContexts.single(targetId, attackContext))

        enqueueRoutine(context = context, options = options) {
            val directory = actor.actorModel?.model?.getMainBattleAnimationDirectory()
            directory?.getNullableChildRecursivelyAs(animProvider.invoke(), EffectRoutineResource::class)
        }
    }

    fun enqueueSubBattleRoutine(targetId: ActorId, attackContext: AttackContext, options: RoutineOptions, animProvider: () -> DatId) {
        val context = ActorContext(actor.id, targetId, attackContexts = AttackContexts.single(targetId, attackContext))

        enqueueRoutine(context = context, options = options) {
            val directory = actor.actorModel?.model?.getSubBattleAnimationDirectory()
            directory?.getNullableChildRecursivelyAs(animProvider.invoke(), EffectRoutineResource::class)
        }
    }

    fun enqueueItemModelRoutine(itemModelSlot: ItemModelSlot, datId: DatId) {
        val context = ActorContext(actor.id, modelSlot = itemModelSlot)
        enqueueRoutine(datId, context = context, options = RoutineOptions(blocking = false)) {
            actor.actorModel?.model?.getEquipmentModelResource(itemModelSlot)
        }
    }

    private fun enqueueRoutine(routineId: DatId, context: ActorContext, options: RoutineOptions = RoutineOptions(), directoryProvider: () -> DirectoryResource?) {
        val provider = RoutineProvider(
            ready = { directoryProvider.invoke() != null },
            provider = { directoryProvider.invoke()?.getNullableChildRecursivelyAs(routineId, EffectRoutineResource::class) }
        )
        enqueueRoutine(context, provider, options)
    }

    private fun enqueueRoutine(context: ActorContext, routineProvider: RoutineProvider, options: RoutineOptions = RoutineOptions()) {
        val queuedRoutine = QueuedRoutine(context, routineProvider, options)

        if (!options.highPriority) {
            routineQueue += queuedRoutine
            return
        }

        val idx = routineQueue.indexOfFirst { !it.options.highPriority }
        if (idx == -1) { routineQueue += queuedRoutine } else { routineQueue.add(idx, queuedRoutine) }
    }

    private fun isTimedOut(queuedRoutine: QueuedRoutine, timeLimit: Duration?): Boolean {
        if (timeLimit == null) { return false }
        return Fps.framesToSeconds(queuedRoutine.age) > timeLimit
    }

    private fun removeNotReadyRoutineIfNeeded(queuedRoutine: QueuedRoutine) {
        if (!isTimedOut(queuedRoutine, notReadyExpiryDuration)) { return }
        routineQueue.removeFirst()
        executeCallbacks(queuedRoutine)
    }

    private fun removeExpiredEntries() {
        while (routineQueue.isNotEmpty()) {
            val next = routineQueue.firstOrNull() ?: return
            if (!isTimedOut(next, next.options.expiryDuration)) { return }

            routineQueue.removeFirst()
            executeCallbacks(next)
        }
    }

    private fun executeCallbacks(queuedRoutine: QueuedRoutine) {
        queuedRoutine.options.callback?.onComplete()
        queuedRoutine.context.attackContexts.getAll().forEach { it.invokeCallback() }
    }

}